src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,357 @@
+        shape.is_inwards.set(0.0);
+        shape.height.set(DEFAULT_HEIGHT);
+        shape.inner_radius.set(60.0);
+        shape.width.set(DEFAULT_WIDTH);
+
+        // FIXME This is an approximation and needs to be computed exactly to avoid clipping in
+        // edge cases.
+        let bbox = Vector2::new(1.5 * DEFAULT_HEIGHT,1.5 * DEFAULT_HEIGHT);
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+
+/// Port definition. Can be parametrised to be either
+/// an InputPort or OutputPort.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct Port<T:ShapeViewDefinition> {
+        spec     : Specification,
+    pub view     : Rc<component::ShapeView<T>>
+}
+
+impl<T:ShapeViewDefinition> Port<T> {
+
+    /// Constructor.
+    pub fn new(spec:Specification) -> Self {
+        let logger = Logger::new("node");
+        let view   = Rc::new(component::ShapeView::<T>::new(&logger));
+        Self{spec,view}.init()
+    }
+
+    fn init(mut self) -> Self {
+        self.update();
+        self
+    }
+
+    /// Modifies the port specification.
+    pub fn mod_specification<F:FnOnce(&mut Specification)>(&mut self, f:F) {
+        f(&mut self.spec);
+        self.update()
+    }
+
+    /// Update the view with our current Specification.
+    fn update(&mut self) {
+        self.update_sprite();
+        self.view.display_object.update();
+    }
+
+    /// Update the position of the sprite according to the Port specification.
+    /// The position is given along a circle, thus the position and rotation of the sprite
+    /// are tied together, so the Port always point in the right direction.
+    fn update_sprite(&mut self) {
+        let translation_vector = na::Vector3::new(0.0,self.spec.inner_radius,0.0);
+        let rotation_vector = -na::Vector3::new(0.0,0.0,self.spec.location.rad().value);
+        let rotation = na::Rotation3::new(rotation_vector);
+        let translation = rotation * translation_vector;
+
+        let node = &self.view.display_object;
+        node.set_position(translation);
+        node.set_rotation(rotation_vector);
+    }
+
+}
+
+/// A port facing towards the center of its inner circle.
+pub type InputPort = Port<InputPortView>;
+
+/// A port facing away from the center of its inner circle.
+pub type OutputPort = Port<OutputPortView>;
+
+
+impl<'t, T:ShapeViewDefinition> From<&'t Port<T>> for &'t Node {
+    fn from(t:&'t Port<T>) -> Self {
+        &t.view.display_object
+    }
+}
+
+impl<T:ShapeViewDefinition> Drop for Port<T> {
+    fn drop(&mut self) {
+        println!("DROP")
+    }
+}
+
+
+
+// ====================
+// === Port Manager ===
+// ====================
+
+/// Handles creation and layouting of ports around a node.
+/// TODO implement the layouting
+#[derive(Debug,Default)]
+pub struct PortManager {
+    parent       : RefCell<Option<WeakNode>>,
+    input_ports  : RefCell<Vec<InputPort>>,
+    output_ports : RefCell<Vec<OutputPort>>,
```

Comment:
Since the `PortManager` is going to be used from the `NodeData`, which can only be used immutably through the `Node`, all functionality here needs to use interior mutability, too.

Reply:
Ok, thats right!